pkg/controller/jobframework/reconciler.go

```diff
@@ -572,66 +569,77 @@ func (r *JobReconciler) recordAdmissionCheckUpdate(wl *kueue.Workload, job Gener
 	}
 }
 
-// IsAncestorJobManaged checks whether an ancestor job is managed by kueue.
-func (r *JobReconciler) IsAncestorJobManaged(ctx context.Context, jobObj client.Object) (bool, error) {
-	ancestor, err := r.getAncestorJobManagedByKueue(ctx, jobObj)
-	if err != nil {
-		return false, err
-	}
-	return ancestor != nil, nil
-}
-
 // getAncestorWorkload returns the Workload object of the Kueue-managed ancestor job.
-func (r *JobReconciler) getAncestorWorkload(ctx context.Context, jobObj client.Object) (*kueue.Workload, error) {
-	ancestor, err := r.getAncestorJobManagedByKueue(ctx, jobObj)
-	if err != nil || ancestor == nil {
-		return nil, err
+func (r *JobReconciler) getAncestorWorkload(ctx context.Context, ancestorJob client.Object) (*kueue.Workload, error) {
+	if QueueNameForObject(ancestorJob) == "" {
+		return nil, nil
 	}
 	wlList := kueue.WorkloadList{}
-	if err := r.client.List(ctx, &wlList, client.InNamespace(ancestor.GetNamespace()), client.MatchingFields{indexer.OwnerReferenceUID: string(ancestor.GetUID())}); client.IgnoreNotFound(err) != nil {
+	if err := r.client.List(ctx, &wlList, client.InNamespace(ancestorJob.GetNamespace()), client.MatchingFields{indexer.OwnerReferenceUID: string(ancestorJob.GetUID())}); client.IgnoreNotFound(err) != nil {
 		return nil, err
 	}
 	if len(wlList.Items) > 0 {
 		// In theory the job can own multiple Workloads, we cannot do too much about it, maybe log it.
+		ctrl.LoggerFrom(ctx).V(2).Info(
+			"WARNING: The job has multiple associated Workloads.",
+			"job", ancestorJob.GetName(),
+			"workloads", klog.KObjSlice(wlList.Items),
+		)
 		return &wlList.Items[0], nil
 	}
 	return nil, nil
 }
 
-// getAncestorJobManagedByKueue traverses controllerRefs to find an ancestor job that is manged by Kueue (ie, it has a queue-name label).
-func (r *JobReconciler) getAncestorJobManagedByKueue(ctx context.Context, jobObj client.Object) (client.Object, error) {
-	seen := sets.New[types.UID]()
-	currentJob := jobObj
+func GetTopLevelAncestorJobManagedByKueue(ctx context.Context, c client.Client, record record.EventRecorder, jobObj client.Object) (client.Object, error) {
+	var (
+		log         = ctrl.LoggerFrom(ctx)
+		seen        = sets.New[types.UID]()
+		currentObj  = jobObj
+		topLevelJob client.Object
+	)
```

Comment:
nit, can we return to the previous form 
```
seen := sets.New[types.UID]()
currentJob := jobObj
```
just so that we minimize the diff with distractions.